lib/ops/opsservice/shrink.go

```diff
@@ -289,15 +306,28 @@ func (s *site) shrinkOperationStart(opCtx *operationContext) (err error) {
 		}
 	}
 
-	s.reportProgress(opCtx, ops.ProgressEntry{
-		State:      ops.ProgressStateInProgress,
-		Completion: 30,
-		Message:    "removing the node from the cluster",
-	})
-
 	// if the node is online, it needs to leave the serf cluster to
 	// prevent joining back
 	if online {
+		if !force {
```

Comment:
Since we know that drain in a production setting can take a very long time, I wonder if adding this step to shrink would just lead to it failing more often for users. Maybe consider keeping the existing behavior with not doing drain by default, and adding a `--drain` flag to remove/leave commands for users who explicitly want it?

Reply:
So my thought on this was, I think when doing a remove command on a node that is online / running, the expectation would be that the uninstall would do a graceful drain of the node if it can. And maybe it should block / fail if the cluster isn't able to accommodate the request.

In which case I can see it being useful, to have a sort of reverse of the flag, uninstall the node without running the drain, which is what I sort of left `--force` as. It will force remove the node regardless of constraints / failures, so that's where I ended up.

So yea, it's definitely difficult, because it does change the behaviour, which causes problems with expectations. But my suspicion is that if we were writing this from scratch, we'd want to include the drain in the uninstall process, so I sort of ended up on changing current behavior, as annoying as that is.

Reply:
@r0mant is this still a concern?

Reply:
Okay, that's reasonable I think. IIRC we already do drain during rolling upgrade so this wouldn't really be much different.

Reply:
Not to add on the PR as such, what about if I want to do `--force` but want it to attempt a graceful drain first, but am fine with it not being able to. Something like `--force --drain`?

Reply:
My knee jerk thought is why do you want to start with a force remove? Wouldn't that be just to trigger a normal uninstall... and then if that fails to repeat with the force flag?

Reply:
That's fair - Depends on the retry logic, i.e. do you retry the current command or do an "if fail do Y" - no problem with that as such. 

For what I described, in the scope of AWS terminations with lifecycle hooks then the gravity side would have to implement it that way since it would be nice to drain gracefully if at all possible but also handling the overall state would be good (so that etcd/serf/etc aren't in a funky state).

Reply:
IIRC, the uninstall hook today just call's a non-force uninstall. But you're right, we're probably not careful that we retry everything for temporary errors, and the force uninstall doesn't really either, it's just if it encounters a temporary error it continues on. The process itself also isn't re-entrant if gravity-site is interrupted, so a long drain could also lead to a weird failure mode.

Probably part of the question should also be what should the behaviour be if the drain fails (IE a pod disruption budget can't be met by the cluster). IIRC the life cycle hook can only keep the node up for some amount of time before it's terminated regardless of the keepalives.

Reply:
Agreed, lots to think about. My inclination is that a PDB should be respected but at some point the node gots to go, even if AWS terminates it, even if it means waiting out the max time. Kubernetes should be doing its best to accommodate the change, especially if you have something like `cluster-autoscaler` installed as well.

Perhaps when this lands in beta/stable https://github.com/kubernetes/enhancements/issues/2000 this will be less of an issue since the kubelet gets more involved and is handling draining, and such.

On the `gravity-site` issue is that due to loss of connectivity or what role does the interruption play here?

Reply:
the uninstall operation isn't checkpointed and the state isn't saved, so if the process is killed off and a new one starts, the new one doesn't look for open / running operations to continue to work on.